# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Dilip-Git18/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

# 1. My Lane (or Freestyle) and Why

## Selected Lane
Predict content that is likely to decline in search performance.

### Why I selected this lane

I selected the content decline prediction lane because identifying pages that are likely to lose search performance can help SEO teams prioritize their work. Instead of reviewing thousands of pages manually, a predictive model can identify pages that may require updates before significant traffic loss occurs.

The FlyRank starter dataset contains historical content performance metrics such as clicks, impressions, CTR, average position, engagement rate, trend percentage, content type, and word count. These variables provide useful signals that may help identify pages at risk of declining performance.

The goal of this project is to build a decision-support system that helps prioritize content updates. The final recommendation should support human decision-making rather than replace it.

# 2. The Question: Decision, Action, Cost of a Wrong Call

## Research Question

Can historical search and engagement metrics be used to identify content pages that are likely to decline in search performance?

## Unit of Analysis

One content page.

## Expected Output

A binary prediction:

- Declining
- Not Declining

## Decision

Which pages should an SEO team review first?

## Action

Pages predicted to decline can be prioritized for content updates, keyword optimization, or technical improvements.

## Cost of an Incorrect Recommendation

### False Positive

A page predicted to decline may actually remain stable, causing unnecessary effort.

### False Negative

A page that is actually declining may not be reviewed in time, potentially leading to reduced traffic and missed opportunities.

Machine learning can assist in prioritizing work but should be used only as decision support rather than making automatic decisions.

In [16]:
import pandas as pd
from pathlib import Path

possible_paths = [
    Path("/content/flyrank-ml-internship/data/raw/content_refresh_anonymized.csv"),
    Path("/content/flyrank-ml-internship/content_refresh_anonymized.csv"),
    Path("../../data/raw/content_refresh_anonymized.csv"),
    Path("../data/raw/content_refresh_anonymized.csv"),
    Path("data/raw/content_refresh_anonymized.csv"),
    Path("content_refresh_anonymized.csv"),
]

dataset_path = None

for path in possible_paths:
    if path.exists():
        dataset_path = path
        break

if dataset_path is None:
    raise FileNotFoundError("Dataset not found!")

print("Dataset found at:", dataset_path)

df = pd.read_csv(dataset_path)

print("Dataset Shape:", df.shape)

display(df.head())

print("\nAverage CTR:", round(df["ctr"].mean(), 2))
print("Average Search Position:", round(df["avg_position"].replace(0, pd.NA).dropna().mean(), 2))
print("Average Engagement Rate:", round(df["engagement_rate"].mean(), 2))

print("\nRows:", len(df))
print("Columns:", len(df.columns))

Dataset found at: /content/flyrank-ml-internship/data/raw/content_refresh_anonymized.csv
Dataset Shape: (30000, 44)


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7



Average CTR: 0.51
Average Search Position: 17.03
Average Engagement Rate: 2.53

Rows: 30000
Columns: 44


## 3. Quick Look at the Data

The starter dataset contains **30,000** content pages and **44** columns, providing a diverse set of SEO and engagement metrics.

The target variable (`is_declining_label`) contains both declining and non-declining pages, making the dataset suitable for a supervised classification problem.

The average CTR is **0.51%**, the average search position is **17.03**, and the average engagement rate is **2.53%**. These metrics indicate variation in page performance, suggesting that historical search and engagement metrics may contain useful signals for predicting future content decline.

These observations support the choice of the content decline prediction lane for the capstone project.

# 4. Careful Words: What I Can and Can't Claim

This project analyzes patterns that are observed in the provided FlyRank dataset.

The machine learning model may identify relationships between historical content metrics and future content decline. These relationships are observational and should be interpreted as decision-support evidence.

This project does not prove that one variable causes another. It also cannot predict Google's ranking algorithm or guarantee future search performance.

The results should be used to help prioritize content review and optimization, while final decisions remain with human experts.

In [17]:
print("Notebook completed successfully.")
print("All sections have been filled.")

Notebook completed successfully.
All sections have been filled.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.